contract/models/contract_line.py

```diff
@@ -0,0 +1,842 @@
+                else:
+                    delay = (date_end - date_start) + timedelta(days=1)
+                rec.delay(delay)
+                contract_line |= rec
+            else:
+                if rec.date_end and rec.date_end < date_start:
+                    rec.stop(date_start)
+                elif (
+                    rec.date_end
+                    and rec.date_end > date_start
+                    and rec.date_end < date_end
+                ):
+                    new_date_start = date_end
+                    new_date_end = date_end + (rec.date_end - date_start)
+                    rec.stop(date_start)
+                    contract_line |= rec.plan_successor(
+                        new_date_start, new_date_end, is_auto_renew
+                    )
+                else:
+                    new_date_start = date_end
+                    new_date_end = (
+                        rec.date_end
+                        if not rec.date_end
+                        else rec.date_end + (date_end - date_start)
+                    )
+                    rec.stop(date_start)
+                    contract_line |= rec.plan_successor(
+                        new_date_start, new_date_end, is_auto_renew
+                    )
+
+        return contract_line
+
+    @api.multi
+    def cancel(self):
+        if not all(self.mapped('is_cancel_allowed')):
+            raise ValidationError(_('Cancel not allowed for this line'))
+        for contract in self.mapped('contract_id'):
+            lines = self.filtered(lambda l, c=contract: l.contract_id == c)
+            msg = _(
+                """Contract line canceled: %s"""
+                % "<br/>- ".join(
+                    [
+                        "<strong>%s</strong>" % name
+                        for name in lines.mapped('name')
+                    ]
+                )
+            )
+            contract.message_post(body=msg)
+        self.mapped('predecessor_contract_line_id').write(
+            {'successor_contract_line_id': False}
+        )
+        return self.write({'is_canceled': True})
+
+    @api.multi
+    def uncancel(self, recurring_next_date):
+        if not all(self.mapped('is_un_cancel_allowed')):
+            raise ValidationError(_('Un-cancel not allowed for this line'))
+        for contract in self.mapped('contract_id'):
+            lines = self.filtered(lambda l, c=contract: l.contract_id == c)
+            msg = _(
+                """Contract line Un-canceled: %s"""
+                % "<br/>- ".join(
+                    [
+                        "<strong>%s</strong>" % name
+                        for name in lines.mapped('name')
+                    ]
+                )
+            )
+            contract.message_post(body=msg)
+        for rec in self:
+            if rec.predecessor_contract_line_id:
+                rec.predecessor_contract_line_id.successor_contract_line_id = (
+                    rec
+                )
```

Comment:
add `assert not  rec.predecessor_contract_line_id.successor_contract_line_id`

Reply:
It's handled by is_un_cancel_allowed

Reply:
Yes, but it's hidden out of view. The assert here would help as a kind of documentation + safety net to make extra sure we do not overwrite the successor of the predecessor if one has been set in the meantime